### PR TITLE
Using Uri newtype, added more Param Types

### PR DESCRIPTION
**Pull request by wz1000**

### CODE REVIEW

File: hie-base/Haskell/Ide/Engine/PluginTypes.hs [hunk 1]
```diff
@@ -18,8 +18,14 @@ module Haskell.Ide.Engine.PluginTypes
   , ParamValP(..)
   , ParamMap
   , pattern ParamTextP
+  , pattern ParamIntP
+  , pattern ParamBoolP
   , pattern ParamFileP
```
Comment (alanz) [inline hie-base/Haskell/Ide/Engine/PluginTypes.hs]: We should possibly remove ParamFileP and ParamPosP, as superseded by the LSP types

File: hie-plugin-api/Haskell/Ide/Engine/PluginUtils.hs [hunk 2]
```diff
@@ -31,8 +32,18 @@ import           Prelude hiding (log)
...
+    Just file -> f file
+    Nothing -> return $ IdeResponseFail (IdeError PluginError
+                 (name <> "Couldn't resolve uri" <> getUri uri) Null)
 
--- |If all the listed params are present in the request resturn their values,
+----------------------------------------
+
+-- |If all the listed params are present in the request return their values,
 -- else return an error message.
 getParams :: (ValidResponse r) =>
```
Comment (alanz) [inline hie-plugin-api/Haskell/Ide/Engine/PluginUtils.hs]: Has the test for this been extended too?

File: hie-apply-refact/Haskell/Ide/ApplyRefactPlugin.hs [hunk 3]
```diff
@@ -87,12 +90,12 @@ lintCmd :: CommandFunc FileDiagnostics
...
     Left err -> return err
-    Right (ParamFile file :& RNil) -> do
-      res <- liftIO $ runEitherT $ runLintCmd (T.unpack file) []
+    Right (ParamFile uri :& RNil) -> pluginGetFile "applyAll: " uri $ \file -> do
+      res <- liftIO $ runEitherT $ runLintCmd file []
       logm $ "lint:res=" ++ show res
       case res of
-        Left diags -> return (IdeResponseOk (FileDiagnostics ("file://" ++ T.unpack file) diags))
-        Right fs   -> return (IdeResponseOk (FileDiagnostics ("file://" ++ T.unpack file) (map hintToDiagnostic fs)))
+        Left diags -> return (IdeResponseOk (FileDiagnostics ("file://" ++ file) diags))
```
Comment (alanz) [inline hie-apply-refact/Haskell/Ide/ApplyRefactPlugin.hs]: J.filePathToUri?

File: hie-base/Haskell/Ide/Engine/PluginTypes.hs (hunk 1 above)
Comment (wz1000) [inline hie-base/Haskell/Ide/Engine/PluginTypes.hs]: ParamFileP and ParmaPosP are using the LSP types under the hood, and are used for passing parameters to the requests.

File: hie-plugin-api/Haskell/Ide/Engine/PluginUtils.hs (hunk 2 above)
Comment (wz1000) [inline hie-plugin-api/Haskell/Ide/Engine/PluginUtils.hs]: Where is the test?

Comment (alanz) [inline hie-plugin-api/Haskell/Ide/Engine/PluginUtils.hs]: See DispatcherSpec.hs. "checking extra params", "reports mismatched params", "reports mismatched optional params" (at least)

File: hie-base/Haskell/Ide/Engine/PluginTypes.hs (hunk 1 above)
Comment (alanz) [inline hie-base/Haskell/Ide/Engine/PluginTypes.hs]: Fair enough, I just wonder if we need multiple position and file types at all. I could go either way, but simplicity is good.

Comment (wz1000) [inline hie-base/Haskell/Ide/Engine/PluginTypes.hs]: There is only one position type, and I plan to convert all the remaining FilePaths to Iris. We need these to signify that files and positions are allowed parameters to the plugins

File: hie-apply-refact/Haskell/Ide/ApplyRefactPlugin.hs (hunk 3 above)
Comment (wz1000) [inline hie-apply-refact/Haskell/Ide/ApplyRefactPlugin.hs]: Yes, I will try to change all of these to Uri

File: hie-base/Haskell/Ide/Engine/PluginTypes.hs (hunk 1 above)
Comment (alanz) [inline hie-base/Haskell/Ide/Engine/PluginTypes.hs]: ok, gotcha

File: hie-plugin-api/Haskell/Ide/Engine/PluginUtils.hs [hunk 4]
```diff
@@ -119,21 +120,46 @@ incorrectParameter name expected value = IdeResponseFail
...
+
+
+    calcRange fm = J.Range s e
+      where
+        sl = fst $ lrNumbers fm
+        sc = 0
+        s = J.Position (sl - 1) sc -- Note: zero-based lines
+        el = snd $ lrNumbers fm
+        ec = length $ last $ lrContents fm
+        e = J.Position (el - 1) ec  -- Note: zero-based lines
```
Comment (alanz) [inline hie-plugin-api/Haskell/Ide/Engine/PluginUtils.hs]: I presume this code came over from `LspStdio.hs`

File: src/Haskell/Ide/Engine/Transport/LspStdio.hs [hunk 5]
```diff
@@ -635,47 +636,8 @@ responseHandlerCb _rin _lf resp = do
...
-
-    calcRange fm = J.Range s e
-      where
-        sl = fst $ lrNumbers fm
-        sc = 0
-        s = J.Position (sl - 1) sc -- Note: zero-based lines
-        el = snd $ lrNumbers fm
-        ec = length $ last $ lrContents fm
-        e = J.Position (el - 1) ec  -- Note: zero-based lines
 {-
```
Comment (alanz) [inline src/Haskell/Ide/Engine/Transport/LspStdio.hs]: Yes

File: test/HaRePluginSpec.hs [hunk 6]
```diff
@@ -64,16 +67,11 @@ hareSpec = do
...
-                                                            "< foo x = x + 3\n"++
-                                                            "---\n"++
-                                                            "> foolong :: Int -> Int\n"++
-                                                            "> foolong x = x + 3\n")
-                                                           ])))
+        Just (IdeResponseOk $
+              jsWrite
+              $ RefactorResult
+              [WorkspaceEdit
+              {_changes = Just (H.fromList [(filePathToUri $ cwd </> "test/testdata/HaReRename.hs",List [TextEdit {_range = Range {_start = Position {_line = 3, _character = 0}, _end = Position {_line = 4, _character = 13}}, _newText = "foolong :: Int -> Int\nfoolong x = x + 3"}])]), _documentChanges = Nothing}] )
```
Comment (alanz) [inline test/HaRePluginSpec.hs]: Please format this and the other test results to not only be on one line, and that it is reasonably clear what is going on, as was done with the original.

**Pull request merged**